pkg/raft/node.go

```diff
@@ -0,0 +1,378 @@
+				return nil
+			}
+		case <-timeoutTicker.C:
+			return errors.New("max wait time reached")
+		}
+	}
+}
+
+func (n *Node) Stop() error {
+	if n == nil {
+		return nil
+	}
+	return n.raft.Shutdown().Error()
+}
+
+// IsLeader returns true if this node is the raft leader
+func (n *Node) IsLeader() bool {
+	if n == nil || n.raft == nil {
+		return false
+	}
+	return n.raft.State() == raft.Leader
+}
+
+func (n *Node) NodeID() string {
+	return n.config.NodeID
+}
+
+func (n *Node) leaderID() string {
+	_, id := n.raft.LeaderWithID()
+	return string(id)
+}
+
+func (n *Node) leaderCh() <-chan bool {
+	return n.raft.LeaderCh()
+}
+
+func (n *Node) leadershipTransfer() error {
+	return n.raft.LeadershipTransfer().Error()
+}
+
+func (n *Node) Config() Config {
+	return *n.config
+}
+
+// Broadcast proposes a block state to be replicated via raft
+func (n *Node) Broadcast(_ context.Context, state *RaftBlockState) error {
+	if !n.IsLeader() {
+		return fmt.Errorf("not leader")
+	}
+
+	data, err := json.Marshal(state) // todo:use protobuf
+	if err != nil {
+		return fmt.Errorf("marshal block state: %w", err)
+	}
+
+	future := n.raft.Apply(data, n.config.SendTimeout)
+	if err := future.Error(); err != nil {
+		return fmt.Errorf("apply log: %w", err)
+	}
+
+	return nil
+}
+
+// GetState returns the current replicated state
+func (n *Node) GetState() RaftBlockState {
+	return *n.fsm.state.Load()
+}
+
+// AddPeer adds a peer to the raft cluster
+func (n *Node) AddPeer(nodeID, addr string) error {
+	n.logger.Debug().Msgf("received join request for remote node %s at %s", nodeID, addr)
+
+	configFuture := n.raft.GetConfiguration()
+	if err := configFuture.Error(); err != nil {
+		return err
+	}
+
+	// remove first when node is already a member of the cluster
+	for _, srv := range configFuture.Configuration().Servers {
+		if srv.ID == raft.ServerID(nodeID) || srv.Address == raft.ServerAddress(addr) {
+			if srv.Address == raft.ServerAddress(addr) && srv.ID == raft.ServerID(nodeID) {
+				n.logger.Debug().Msgf("node %s at %s already member of cluster, ignoring join request", nodeID, addr)
+				return nil
+			}
+			future := n.raft.RemoveServer(srv.ID, 0, 0)
+			if err := future.Error(); err != nil {
+				return fmt.Errorf("removing existing node %s at %s: %w", nodeID, addr, err)
+			}
+		}
+	}
+
+	f := n.raft.AddVoter(raft.ServerID(nodeID), raft.ServerAddress(addr), 0, 0)
+	if f.Error() != nil {
+		return f.Error()
+	}
+	n.logger.Debug().Msgf("node %s at %s joined successfully", nodeID, addr)
+	return nil
+}
+
+// RemovePeer removes a peer from the raft cluster
+func (n *Node) RemovePeer(nodeID string) error {
+	future := n.raft.RemoveServer(raft.ServerID(nodeID), 0, 0)
+	return future.Error()
+}
+
+// Shutdown stops the raft node
+func (n *Node) Shutdown() error {
+	return n.raft.Shutdown().Error()
+}
+
+// SetApplyCallback sets a callback to be called when log entries are applied
+func (n *Node) SetApplyCallback(ch chan<- RaftApplyMsg) {
```

Comment:
nit: what is this for? the go doc is very light

Reply:
The channel is passed by the syncer to receive first level state updates from within the raft cluster. This should be the fastest communication channel available.